### PR TITLE
[WIP] [Feedback needed] Visualization Fixes #1

### PR DESCRIPTION
**Pull request by marcusnovotny**
Not ready to merge yet. Looking for feedback.

Working on #71. This push fixes:

> Start on time-range with data. Prevent zooming / panning out of our data-range.

I've also worked on the episode visualization, but it's not done yet. So far I've implemented **two different approaches:**
- Adding "bars" to the episode airing dates. Currently only 1 px wide. Those could be made wider & equipped with a click event handler, so they display a tooltip with the episode code & title on clicking. 
  - Pro: Eliminates issues with overlying episode codes.
  - Con: Episode data can only be seen after clicking on the bar - And then only this episode.
- Adding a seperate x-axis only for the episode dates.
  - Pro: Displays all episode data immediately.
  - Con: Needs some multi-scale formatting to prevent overlying text (not sure how to implement yet)

Please have a look and tell me what solution you'd prefer / if there's other ways that I am currently missing.
### Known issues:
- Resizing to a bigger window does not work properly yet
- There's some "double episodes" with the same airing date. The episode bars have a stronger color there. I'd need to combine those episodes somehow to show them in one tooltip / on one axis tick.
- Episode axis labels ignore container boundaries. Not sure why, will have to implement a clip-path there
- A few axis labels are out of order (again not sure why, but I can just sort them)
- Overlying text on the episode axis. Perfect way to solve this would be multi-scale formatting with 3 levels:
  - Seasons only for multiple-year-view (e.g. Season 1, Season 2, Season 3)
  - Code only for multiple months (e.g. S01E01, S01E02, S01E03, S01E04, S01E05...)
  - Full title for less max three episodes on the axis
    ... not sure how to implement that though.

Looking forward to your suggestions!

### CODE REVIEW

File: example/app/views/chart.html [hunk 1]
```diff
@@ -10,7 +10,7 @@
     <script src="/assets/d3.min.js"></script>
     <script src="/chart.js"></script>
     <script>
-    var chart = characterChart(d3.select("#chart"), "/csv/{{slug}}.csv");
+    var chart = characterChart(d3.select("#chart"), "/csv/{{slug}}.csv", "/csv/episodes.csv");
```
Comment (julienschmidt) [inline example/app/views/chart.html:13]: The function receives a path like `/very/custom/path/Donald_Trump.csv`. Then you remove the filename and  `/very/custom/path/episodes.csv` contains the episodes, `/very/custom/path/Donald_Trump/2016-03.csv` the data for this month etc.

File: public/chart.js [hunk 2]
```diff
@@ -111,6 +128,15 @@ function characterChart(svg, dataURL) {
         };
     }
 
+    function convertEpisodesCSV(d) {
+        // convert string data from CSV
+        return {
+            date: parseDate(d.date.substring(0, 10)),
```
Comment (julienschmidt) [inline public/chart.js:134]: Fixed it server-side already. No need for the substring anymore (`.slice()` is actually what you are looking for `.substring()` and `.substr()` are both somehow broken)

File: public/chart.js [hunk 3]
```diff
@@ -161,18 +227,39 @@ function characterChart(svg, dataURL) {
...
 
-    // request csv data and fill chart
-    d3.csv(dataURL, convertCSV, fillChart);
+    // request csv twitter data and fill chart
+    d3.csv(dataURL, convertTwitterCSV, fillChart);
+
+    // request csv episodes data and create episode labels
+    d3.csv(episodesURL, convertEpisodesCSV, episodeLabels)
 
     this.resize = render;
```
Comment (julienschmidt) [inline public/chart.js:264]: please replace `this` with `var whatever = {}; whatever.resize = ...`. This refers to the global scope here. Mea culpa

**Comment by julienschmidt** [comment]
Run `npm test` before committing to run a JS linter which lets you know about all [the semicolons you forgot](https://travis-ci.org/Rostlab/JS16_ProjectD_Group4/jobs/118873428) :wink:

**Comment by marcusnovotny** [comment]
Thanks for the feedback! :) Any thoughts about how to implement the episode labels best? I'm a little stuck on that one

**Comment by julienschmidt** [comment]
I wasn't able to take a look at your visual changes yet since the crawler is still searching for tweets mentioning Jon Snow... :scream: 

But maybe show the episodes as a "shadow" in the background as wide as the day when it aired? After all we have no specific time or runtime. And then there are different time zones, VoD viewers (and torrent etc). I think it is fair to highlight the whole day. 
But I think you have to shift all dates +1 since it should be already the next day in UTC.

File: public/chart.js (hunk 3 above)
Comment (marcusnovotny) [inline public/chart.js:264]: I'm using this in line 5 and following aswell, does that have to be changed too?

Comment (julienschmidt) [inline public/chart.js:264]: Yes. Don't use `this` within this function. Rather generate a new object and attach properties to it as I described above.

Comment (marcusnovotny) [inline public/chart.js:264]: Your proposed code doesn't work. Doesn't "this" refer the characterChart function object?

Comment (julienschmidt) [inline public/chart.js:264]: there is no `characterChart` object until you create it with `new ...`

I'll take a look.

Comment (marcusnovotny) [inline public/chart.js:264]: What if we do just that in chart.html?

`var chart = new characterChart(d3.select("#chart"), "/csv/{{slug}}.csv");`

Then the this in characterChart actually references the function

EDIT: Works fine for me

**Comment by marcusnovotny** [comment]
Found a nice workaround for the episode labels, will push later

**Pull request closed without merging**